platform-sdk/swirlds-virtualmap/src/main/java/com/swirlds/virtualmap/internal/cache/VirtualNodeCache.java

```diff
@@ -743,26 +733,36 @@ public VirtualLeafRecord<K, V> lookupLeafByPath(final long path, final boolean f
      * 		The last leaf path to receive in the results. It is possible, through merging of multiple rounds,
      * 		for the data to have leaf data that is outside the expected range for the {@link VirtualMap} of
      * 		this cache. We need to provide the leaf boundaries to compensate for this.
-     * @return A non-null stream of dirty leaves. May be empty. Will not contain duplicate records. Will be sorted
-     * 		by path, with the lowest (root-most) path first.
+     * @param sorted
+     *      Indicates if the resulting stream should be sorted by path or not
+     * @return A non-null stream of dirty leaves. May be empty. Will not contain duplicate records
      * @throws MutabilityException
      * 		if called on a cache that still allows dirty leaves to be added
      */
-    public Stream<VirtualLeafRecord<K, V>> dirtyLeaves(final long firstLeafPath, final long lastLeafPath) {
+    public Stream<VirtualLeafRecord<K, V>> dirtyLeaves(
+            final long firstLeafPath, final long lastLeafPath, final boolean sorted) {
         if (!dirtyLeaves.isImmutable()) {
             throw new MutabilityException("Cannot call on a cache that is still mutable for dirty leaves");
         }
-
-        final AtomicReference<Mutation<K, VirtualLeafRecord<K, V>>> lastSeen = new AtomicReference<>();
-        return dirtyLeaves
-                .sortedStream(dirtyLeafComparator)
+        final Stream<VirtualLeafRecord<K, V>> result = dirtyLeaves.stream()
                 .filter(mutation -> {
                     final long path = mutation.value.getPath();
                     return path >= firstLeafPath && path <= lastLeafPath;
                 })
-                .filter(mutation -> dedupeLeafByPath(mutation, lastSeen))
-                .filter(mutation -> !mutation.deleted)
+                .filter(mutation -> {
+                    if (mutation.next != null) {
+                        mutation.next.setFiltered();
+                    }
+                    return true;
+                })
+                .filter(mutation -> !mutation.isFiltered())
```

Comment:
You can't use `setFiltered()` and `isFiltered()` is one pass: a record may be read from the stream and later be marked as filtered.

Reply:
Should it be possible with reverse stream over the array?

Reply:
OK, it's now done in two passes. First, all obsolete mutations are explicitly marked to filter, using `ConcurrentArray.parallelTraverse()` mechanism. Second, the resulting stream is filtered based on the marks.